### PR TITLE
[Issue 4023] Rework config file format

### PR DESCRIPTION
**Pull request by mbaxter**
## PR Description
Rework config files and config parsing to match new format.  Also fix some issues where `InputStream`'s were not being closed.

Remaining todos include moving config values that are now constants out of `SpecConfig`.  

## Fixed Issue(s)
#4023 

## Documentation

The network config  (`--network`) can no longer point to a directory (containing `phase0.yaml` and `altair.yaml` files).  We will only parse singular yaml files.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

### CODE REVIEW

**Review by ajsutton — COMMENTED** [review]
We've wound up losing some values from the `/eth/v1/config/spec` endpoint which we probably need to keep for backwards compatibility.

Removed by this change:
```
   "DOMAIN_AGGREGATE_AND_PROOF": "0x06000000",
    "DOMAIN_BEACON_ATTESTER": "0x01000000",
    "DOMAIN_BEACON_PROPOSER": "0x00000000",
    "DOMAIN_CONTRIBUTION_AND_PROOF": "0x09000000",
    "DOMAIN_DEPOSIT": "0x03000000",
    "DOMAIN_RANDAO": "0x02000000",
    "DOMAIN_SELECTION_PROOF": "0x05000000",
    "DOMAIN_SYNC_COMMITTEE": "0x07000000",
    "DOMAIN_SYNC_COMMITTEE_SELECTION_PROOF": "0x08000000",
    "DOMAIN_VOLUNTARY_EXIT": "0x04000000",
    "EPOCHS_PER_RANDOM_SUBNET_SUBSCRIPTION": "256",
    "RANDOM_SUBNETS_PER_VALIDATOR": "1",
    "TARGET_AGGREGATORS_PER_COMMITTEE": "16",
    "TARGET_AGGREGATORS_PER_SYNC_SUBCOMMITTEE": "4",
```

We also picked up some values from the sharing and merge forks which we probably don't want to include yet as the names are likely to change:

```
    "MERGE_FORK_EPOCH": "18446744073709551615",
    "MERGE_FORK_VERSION": "0x02000000",
    "SHARDING_FORK_EPOCH": "18446744073709551615",
    "SHARDING_FORK_VERSION": "0x03000000",
    "TRANSITION_TOTAL_DIFFICULTY": "4294967296",
```

`PRESET_BASE` is also new but I think that's a good thing to add.

File: util/src/main/java/tech/pegasys/teku/util/config/ConstantsReader.java [hunk 1]
```diff
@@ -15,18 +15,55 @@
...
 import java.util.Map;
+import java.util.Optional;
 import java.util.function.Function;
+import java.util.stream.Collectors;
 import org.apache.tuweni.bytes.Bytes;
+import tech.pegasys.teku.infrastructure.io.resource.ResourceLoader;
 import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 import tech.pegasys.teku.ssz.type.Bytes4;
 
 class ConstantsReader {
```
Comment (ajsutton) [inline util/src/main/java/tech/pegasys/teku/util/config/ConstantsReader.java:36]: Not for this PR, but I'm beginning to wonder if we should change this to just explicitly set the values in `Constants` from a given `SpecConfig`.  We know there won't be any new fields added to this class so the magic reflection isn't needed and then we could avoid having duplicate code for reading the config files.

**Review by ajsutton — COMMENTED** [review]
Code looks good, we just need to come up with a plan for the `/config/specs` endpoint.

**Comment by ajsutton** [comment]
> we just need to come up with a plan for the /config/specs endpoint.

Looking at this quickly, we're just passing a map with the `rawConfig` which is what gets returned, so we could shove the extra constants in there manually.

We could also remove the values from future forks by removing anything in the ignore list from the raw map as well.  The advantage of that is that if we start using those values we'd have to remove them from the ignore list and they'll then automatically appear in the spec output.

**Comment by mbaxter** [comment]
@ajsutton - I ended up manually setting the missing constants in the REST API's `ConfigProvider`.  I think we can just cut that custom logic once all of the clients support the new format.  

And I have bumped up the logic that ignores the future-fork-related config values so that we don't save those in the `rawConfig` and therefore don't display them from the API.  That includes the new `PRESET_BASE` key - I could add that back in, but its cleaner to just treat all of the ignored keys the same way and the preset-related values are already flattened into the config, WDYT?

**Review by ajsutton — APPROVED** [review]
LGTM.

**Comment by ajsutton** [comment]
> And I have bumped up the logic that ignores the future-fork-related config values so that we don't save those in the rawConfig and therefore don't display them from the API. That includes the new PRESET_BASE key - I could add that back in, but its cleaner to just treat all of the ignored keys the same way and the preset-related values are already flattened into the config, WDYT?

Yep, I think that makes sense to me.  We may want to put the PRESET_BASE back in at some point, but I'm happy to leave it out for now so this change doesn't affect the API at all.

**Comment by bgravenorst** [comment]
Doc changes for this has already been implemented, see https://docs.teku.pegasys.tech/en/latest/Reference/CLI/CLI-Syntax/#network

**Pull request merged**